Propensity score Joint Estimation Versus 2 stage Estimation

NathanielF opened the pull request:
Working on a draft PR to improve or augment the existing propensity scoring weighting implementation. In particular to make it a bit faster and more "Bayesian". As it currently stands we are performing a two-step strategy where we fit a propensity score model and then push the values of the posterior estimate for the propensity score through a re-weighting routine to estimate the causal contrast. 

But we could try and explore a more properly Bayesian model where we fit the propensity score outcome and the model outcome at once in the same model context. This more properly Bayesian and a good bit faster. 

See for instance, work here: https://github.com/ajnafa/Latent-Bayesian-MSM by Jordan Nafa and Andrew Heiss

---- EDIT ----

So i think i've finally understood the role of the propensity score in Bayesian estimation and I demonstrate that the two-stage method is what we actually want. The changes i've made are actually quite small. I've added a function to the PyMC model associated with the inverse propensity experiment class. 

But in the notebook example i show why this two-stage process is to be preferred over the joint fit model. 



<!-- readthedocs-preview causalpy start -->
----
📚 Documentation preview 📚: https://causalpy--500.org.readthedocs.build/en/500/

<!-- readthedocs-preview causalpy end -->

juanitorduz commented:
@NathanielF I love these examples 💪 ! Maybe you can ask Andrew Heiss or Jordan Nafa for feedback (maybe once we merge it so that this is not a blocker?)

juanitorduz commented:
I left some minor comments about adding more explanatory tests is some sections.

NathanielF commented:
Thanks @juanitorduz ! Good idea to ask them. Appreciate the quick feedback!!!

NathanielF commented:
<p>Honestly it's super annoying. Think related to a bug or mismatch between pymc and apple mac: <a href="https://discourse.pymc.io/t/dataset-size-dependent-eoferror/11977/14" target="_blank">https://discourse.pymc.io/t/dataset-size-dependent-eoferror/11977/14</a></p>

---

 <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/500/discussion/#comment-3080140183'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Done. </p>

---

 <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/500/discussion/#comment-3080140262'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Not sure i remember how to link section headings?</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/500/discussion/#comment-3080140348'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>I left this as it was </p>

---

 <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/500/discussion/#comment-3080140469'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>added a box to this effect</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/500/discussion/#comment-3080140578'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>added a box to highlight. </p><p> </p><p>Re: IV models. I think it's different the equations aren't directly influencing one another. The outcomes of the two equations in the IV are being modelled in the likelihood as a bivariate potentially correlated pair. I think an instrument plays a different role than a propensity score and the modularity isn't required in the same way. </p>

---

 <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/500/discussion/#comment-3080141009'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>renamed</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/500/discussion/#comment-3080141148'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>I.E. one prediction doesn't feed into the next, we just allow that there maybe some correlation between both predictions</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/500/discussion/#comment-3080141009'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>I've added the Fan li reference throughout. It's mainly her talk and paper that i'm referring to here. Added again in the conclusion, rather than here because it's more likely to spur follow up there. </p>

---

 <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/500/discussion/#comment-3080141412'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Added </p>

---

 <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/500/discussion/#comment-3080141518'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>used item accessor to tidy</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/500/discussion/#comment-3080141603'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>used item accessor to tidy</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/500/discussion/#comment-3080141726'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Fixed</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/500/discussion/#comment-3080141956'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>hidden</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/500/discussion/#comment-3080142069'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>added. </p>

---

 <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/500/discussion/#comment-3080142258'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>added more citations</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/500/discussion/#comment-3080142657'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>hidden</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/500/discussion/#comment-3080142527'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>added a spline plot and admonition box explaining the idea</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/500/discussion/#comment-3080140913'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

drbenvincent commented:
<p>Weird, I don't need to do that on my Mac. No matter</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/500/discussion/#comment-3080140183'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>
